contracts/royalty-enforcement/RoyaltyAllowlist.sol

```diff
@@ -0,0 +1,160 @@
+// Access Control
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+// Introspection
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+// Interfaces
+import "./IRoyaltyAllowlist.sol";
+
+// Interface to retrieve the implemention stored inside the Proxy contract
+interface IProxy {
+    // Returns the current implementation address used by the proxy contract
+    function PROXY_getImplementation() external view returns (address);
+}
+
+/*
+    RoyaltyAllowlist is an implementation of a Allowlist registry, storing addresses and bytecode
+    which are allowed to be approved operators and execute transfers of interfacing token contracts (e.g. ERC721/ERC1155). The registry
+    will be a deployed contract that tokens may interface with and point to.
+*/
+contract RoyaltyAllowlist is ERC165, AccessControl, IRoyaltyAllowlist {
+    ///     =====       Events       =====
+
+    /// @dev Emitted when a target address is added or removed from the Allowlist
+    event AddressAllowlistChanged(address indexed target, bool added);
+
+    /// @dev Emitted when a target smart contract wallet is added or removed from the Allowlist
+    event WalletAllowlistChanged(
+        bytes32 indexed targetBytes,
+        address indexed targetAddress,
+        bool added
+    );
+
+    ///     =====   State Variables  =====
+
+    /// @dev Only REGISTRAR_ROLE can invoke white listing registration and removal
+    bytes32 public constant REGISTRAR_ROLE = bytes32("REGISTRAR_ROLE");
+
+    /// @dev Mapping of Allowlisted addresses
+    mapping(address => bool) private addressAllowlist;
+
+    /// @dev Mapping of Allowlisted bytecodes
+    mapping(bytes32 => bool) private bytecodeAllowlist;
+
+    ///     =====   Constructor  =====
+
+    /**
+     * @dev Grants `DEFAULT_ADMIN_ROLE` to the supplied `admin` address
+     */
+    constructor(address admin) {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+    }
+
+    ///     =====   View functions  =====
+
+    /// @dev ERC-165 interface support
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override(ERC165, AccessControl) returns (bool) {
+        return
+            interfaceId == type(IRoyaltyAllowlist).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    /// @dev Returns true if an address is Allowlisted, false otherwise
+    function isAllowlisted(
+        address target
+    ) external view override returns (bool) {
+        if (addressAllowlist[target]) {
+            return true;
+        }
+
+        // Check if caller is a Allowlisted smart contract wallet
+        bytes32 codeHash;
+        assembly {
+            codeHash := extcodehash(target)
+        }
+        if (bytecodeAllowlist[codeHash]) {
+            // If wallet proxy bytecode is approved, check addr of implementation contract
+            address impl = IProxy(target).PROXY_getImplementation();
+
+            return addressAllowlist[impl];
+        }
```

Comment:
what about non-smart contract wallet bytecode allowlists, e.g. allowlist the bytecode of a standard ERC20 impl?

Reply:
Curious as to how this relates to royalties, though. I originally thought about leaving the bytecode stuff in but couldn't think of a use-case past SCW's.